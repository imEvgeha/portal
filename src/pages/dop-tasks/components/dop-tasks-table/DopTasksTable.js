import React, {useState, useEffect} from 'react';
import PropTypes from 'prop-types';
import Tag from '@atlaskit/tag/dist/cjs/Tag';
import config from 'react-global-configuration';
import {compose} from 'redux';
import NexusGrid from '../../../../ui/elements/nexus-grid/NexusGrid';
import {GRID_EVENTS} from '../../../../ui/elements/nexus-grid/constants';
import createValueFormatter from '../../../../ui/elements/nexus-grid/elements/value-formatter/createValueFormatter';
import withColumnsResizing from '../../../../ui/elements/nexus-grid/hoc/withColumnsResizing';
import withFilterableColumns from '../../../../ui/elements/nexus-grid/hoc/withFilterableColumns';
import withInfiniteScrolling from '../../../../ui/elements/nexus-grid/hoc/withInfiniteScrolling';
import withSideBar from '../../../../ui/elements/nexus-grid/hoc/withSideBar';
import withSorting from '../../../../ui/elements/nexus-grid/hoc/withSorting';
import {COLUMN_MAPPINGS, USER, INITIAL_SEARCH_PARAMS, DOP_GUIDED_TASK_URL, DOP_PROJECT_URL} from '../../constants';
import {fetchDopTasksData} from '../../utils';
import DopTasksTableStatusBar from '../dop-tasks-table-status-bar/DopTasksTableStatusBar';
import './DopTasksTable.scss';

const DopTasksTableGrid = compose(
    withSideBar(),
    withFilterableColumns(),
    withColumnsResizing(),
    withSorting(),
    withInfiniteScrolling({fetchData: fetchDopTasksData})
)(NexusGrid);

const DopTasksTable = ({user}) => {
    const [paginationData, setPaginationData] = useState({
        pageSize: 0,
        totalCount: 0,
    });
    const [externalFilter, setExternalFilter] = useState({
        user,
    });
    const formattedValueColDefs = COLUMN_MAPPINGS.map(col => {
        if (col.colId === 'taskName') {
            return {
                ...col,
                cellRendererParams: {
                    link: `${config.get('gateway.DOPUrl')}${DOP_GUIDED_TASK_URL}`,
                },
            };
        }
        if (col.colId === 'taskStatus') {
            return {
                ...col,
                cellRendererFramework: params => {
                    const {value} = params || {};
                    let color = 'yellowLight';
                    switch (value) {
                        case 'COMPLETED':
                            color = 'grey';
                            break;
                        case 'READY':
                            color = 'green';
                            break;
                        case 'IN PROGRESS':
                            color = 'blue';
                            break;
                        default:
                            color = 'standard';
                            break;
                    }
                    return (
                        <div>
                            <Tag text={value} color={color} />
                        </div>
                    );
                },
            };
        }
        if (col.colId === 'projectName') {
            return {
                ...col,
                cellRendererParams: {
                    link: `${config.get('gateway.DOPUrl')}${DOP_PROJECT_URL}`,
                    linkId: 'projectId',
                },
            };
        }
        return {
            ...col,
            valueFormatter: createValueFormatter(col),
        };
    });

    useEffect(() => {
        if (externalFilter.user !== user) {
            setExternalFilter(prevData => {
                return {
                    ...prevData,
                    user,
                };
            });
        }
    }, [user]);

    const getPaginationData = ({api}) => {
        const pageSize = api.paginationGetPageSize();
        const totalCount = api.paginationGetRowCount();
        if (totalCount > 0) {
            setPaginationData({pageSize, totalCount});
        }
    };

    const onGridReady = ({type, api}) => {
        const {FILTER_CHANGED} = GRID_EVENTS;
        switch (type) {
            case FILTER_CHANGED: {
                // console.log(api.getFilterModel());
                break;
            }
            default:
                break;
        }
    };

    const onSortChanged = ({api}) => {
        // get sorting column and prepare data for passing it as a payload instead of url params (not supported by DOP api)
        const sortModel = api.getSortModel();
        if (sortModel.length) {
            const sortCriterion = [
                {
                    fieldName: sortModel[0].colId,
                    ascending: sortModel[0].sort === 'asc',
                },
            ];
            setExternalFilter(prevData => {
                return {
                    ...prevData,
                    sortCriterion,
                };
            });
        } else {
            setExternalFilter(prevData => {
                return {
                    ...prevData,
                    sortCriterion: INITIAL_SEARCH_PARAMS.sortCriterion,
                };
            });
        }
    };

    return (
        <div className="nexus-c-dop-tasks-table">
            <DopTasksTableGrid
                id="DopTasksTable"
                columnDefs={formattedValueColDefs}
                mapping={COLUMN_MAPPINGS}
                suppressRowClickSelection
                onGridEvent={onGridReady}
                onSortChanged={onSortChanged}
                pagination={true}
                suppressPaginationPanel={true}
                onPaginationChanged={getPaginationData}
                externalFilter={externalFilter}
            />
            <DopTasksTableStatusBar paginationData={paginationData} />
        </div>
    );
};

DopTasksTable.propTypes = {
    user: PropTypes.string,
};

DopTasksTable.defaultProps = {
    user: USER,
};

export default DopTasksTable;
