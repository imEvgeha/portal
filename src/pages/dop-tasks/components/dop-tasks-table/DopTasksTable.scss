@import '~@vubiquity-nexus/portal-styles/scss/config.scss';
@import '~@vubiquity-nexus/portal-styles/scss/atlaskit/mixins';
@import '~@vubiquity-nexus/portal-styles/scss/atlaskit/colors';
@import '~@portal/portal-components/styles/color-palette';

.nexus-c-dop-tasks-table {
    position: relative;
    display: flex;
    flex-direction: column;
    width: 100%;
    height: 100%;
    @include padding(0, 6, 10, 6);

    &__more-actions {
        margin-bottom: 10px;
        cursor: pointer;
        fill: map-get($neutrals, N70);
    }
    &__action-menu {
        position: absolute;
        top: 25px;
        background: map-get($neutrals, N0);
        border: 1px solid map-get($neutrals, N70);
        z-index: 2;
        width: 150px;
        padding: 10px;

        &--item {
            color: map-get($neutrals, N70);
            pointer-events: none;
            white-space: nowrap;
            padding-bottom: 10px;
            &.enable-option {
                pointer-events: all;
                color: map-get($neutrals, N900);
                cursor: pointer;
            }
            &:last-of-type {
                padding-bottom: 0;
            }
        }
    }

    & .nexus-c-nexus-grid {
        height: 100%;
        margin: 0;
        .nexus-c-grid-checkbox--is-disabled {
            .ag-checkbox-input-wrapper {
                background-color: palette(black, 10);
            }
        }
    }
}
