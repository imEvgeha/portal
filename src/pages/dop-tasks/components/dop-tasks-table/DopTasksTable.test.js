import React from 'react';
import {shallow} from 'enzyme';
import DopTasksTable from './DopTasksTable';

describe('DopTasksTable', () => {
    let wrapper = null;

    const props = {
        setExternalFilter: jest.fn(),
        setGridApi: jest.fn(),
        setColumnApi: jest.fn(),
        assignTasks: jest.fn(),
        unAssignTasks: jest.fn(),
        changePriority: jest.fn(),
    };

    beforeEach(() => {
        wrapper = shallow(<DopTasksTable {...props} />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render MoreIcon', () => {
        expect(wrapper.find('.nexus-c-dop-tasks-table__more-actions').length).toEqual(1);
    });
});
