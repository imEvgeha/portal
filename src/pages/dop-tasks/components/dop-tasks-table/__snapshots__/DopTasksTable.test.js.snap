// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DopTasksTable should match snapshot 1`] = `
<div
  className="nexus-c-dop-tasks-table"
>
  <i
    className="po po-more nexus-c-dop-tasks-table__more-actions"
    onClick={[Function]}
  />
  <Component
    columnDefs={Array []}
    externalFilter="user"
    id="DopTasksTable"
    mapping={
      Array [
        Object {
          "cellRenderer": "loadingCellRenderer",
          "colId": "taskName",
          "enableSearch": true,
          "field": "taskName",
          "headerName": "Task Name",
          "javaVariableName": "taskName",
          "minWidth": 200,
          "searchDataType": "string",
        },
        Object {
          "colId": "taskStatus",
          "enableSearch": true,
          "field": "taskStatus",
          "headerName": "Status",
          "javaVariableName": "taskStatus",
          "options": Array [
            "READY",
            "IN PROGRESS",
            "COMPLETED",
            "EXITED",
            "OBSOLETE",
          ],
          "searchDataType": "multiselect",
          "width": 180,
        },
        Object {
          "colId": "activityEstimatedEndDate",
          "dataType": "regionalMidnight",
          "field": "activityEstimatedEndDate",
          "headerName": "Due Date",
          "javaVariableName": "activityEstimatedEndDate",
          "searchDataType": "businessDateTime",
          "width": 120,
        },
        Object {
          "cellRenderer": "loadingCellRenderer",
          "colId": "projectName",
          "enableSearch": true,
          "field": "projectName",
          "headerName": "Project Name",
          "javaVariableName": "projectName",
          "minWidth": 200,
          "searchDataType": "string",
        },
        Object {
          "colId": "customPriority",
          "enableSearch": true,
          "field": "customPriority",
          "headerName": "Priority",
          "javaVariableName": "customPriority",
          "minWidth": 80,
          "options": Array [
            "Low",
            "Medium",
            "High",
          ],
          "searchDataType": "multiselect",
        },
        Object {
          "colId": "OrderExternalID",
          "enableSearch": true,
          "field": "OrderExternalID",
          "headerName": "External Order ID",
          "javaVariableName": "OrderExternalID",
          "searchDataType": "string",
          "width": 150,
        },
        Object {
          "colId": "Customer",
          "enableSearch": true,
          "field": "Customer",
          "headerName": "Customer Name",
          "javaVariableName": "Customer",
          "searchDataType": "string",
          "width": 180,
        },
        Object {
          "colId": "servicingRegion",
          "enableSearch": true,
          "field": "servicingRegion",
          "headerName": "Servicing Region",
          "javaVariableName": "servicingRegion",
          "searchDataType": "string",
          "width": 120,
        },
        Object {
          "colId": "potentialOwners",
          "enableSearch": true,
          "field": "potentialOwners",
          "headerName": "Work Queue",
          "javaVariableName": "potentialOwners",
          "searchDataType": "string",
          "width": 120,
        },
        Object {
          "colId": "actualOwner",
          "enableSearch": true,
          "field": "actualOwner",
          "headerName": "Owner",
          "javaVariableName": "actualOwner",
          "searchDataType": "string",
          "width": 120,
        },
        Object {
          "colId": "activityActualStartDate",
          "dataType": "regionalMidnight",
          "field": "activityActualStartDate",
          "headerName": "Actual Start Date",
          "hide": true,
          "javaVariableName": "activityActualStartDate",
          "searchDataType": "businessDateTime",
          "width": 120,
        },
        Object {
          "colId": "activityActualEndDate",
          "dataType": "regionalMidnight",
          "field": "activityActualEndDate",
          "headerName": "Actual Completion Date",
          "hide": true,
          "javaVariableName": "activityActualEndDate",
          "searchDataType": "businessDateTime",
          "width": 120,
        },
        Object {
          "colId": "activityPlannedCompletionDate",
          "dataType": "regionalMidnight",
          "field": "activityPlannedCompletionDate",
          "headerName": "Planned Due Date",
          "hide": true,
          "javaVariableName": "activityPlannedCompletionDate",
          "searchDataType": "businessDateTime",
          "width": 120,
        },
        Object {
          "colId": "projectStartDate",
          "dataType": "regionalMidnight",
          "field": "projectStartDate",
          "headerName": "Project Start Date",
          "hide": true,
          "javaVariableName": "projectStartDate",
          "searchDataType": "businessDateTime",
          "width": 120,
        },
        Object {
          "colId": "projectPlannedCompletionDate",
          "dataType": "regionalMidnight",
          "field": "projectPlannedCompletionDate",
          "headerName": "Project Planned Due Date",
          "hide": true,
          "javaVariableName": "projectPlannedCompletionDate",
          "searchDataType": "businessDateTime",
          "width": 120,
        },
        Object {
          "colId": "projectStatus",
          "enableSearch": true,
          "field": "projectStatus",
          "headerName": "Project Status",
          "hide": true,
          "javaVariableName": "projectStatus",
          "options": Array [
            "NOT STARTED",
            "IN PROGRESS",
            "RELEASE IN PROGRESS",
            "COMPLETED",
            "CANCEL IN PROGRESS",
            "CANCELLED",
            "HOLD IN PROGRESS",
            "IN ERROR",
          ],
          "searchDataType": "multiselect",
          "width": 120,
        },
        Object {
          "colId": "language",
          "configEndpoint": "/languages",
          "enableSearch": true,
          "field": "language",
          "headerName": "Language",
          "hide": true,
          "javaVariableName": "language",
          "searchDataType": "multiselect",
          "width": 100,
        },
        Object {
          "colId": "locale",
          "configEndpoint": "/countries",
          "enableSearch": true,
          "field": "locale",
          "headerName": "Locale",
          "hide": true,
          "javaVariableName": "locale",
          "searchDataType": "multiselect",
          "width": 100,
        },
        Object {
          "colId": "licenseStartDate",
          "dataType": "regionalMidnight",
          "field": "licenseStartDate",
          "headerName": "License Start Date",
          "hide": true,
          "javaVariableName": "licenseStartDate",
          "searchDataType": "businessDateTime",
          "width": 120,
        },
        Object {
          "colId": "affiliate",
          "configEndpoint": "/affiliates",
          "enableSearch": true,
          "field": "affiliate",
          "headerName": "Affiliates",
          "hide": true,
          "javaVariableName": "affiliate",
          "searchDataType": "multiselect",
          "width": 100,
        },
        Object {
          "colId": "territory",
          "configEndpoint": "/countries",
          "enableSearch": true,
          "field": "territory",
          "headerName": "Territories",
          "hide": true,
          "javaVariableName": "territory",
          "searchDataType": "multiselect",
          "width": 100,
        },
        Object {
          "colId": "contentType",
          "configEndpoint": "/contentType",
          "enableSearch": true,
          "field": "contentType",
          "headerName": "Content Type",
          "hide": true,
          "javaVariableName": "contentType",
          "searchDataType": "multiselect",
          "width": 100,
        },
        Object {
          "colId": "coreTitleID",
          "enableSearch": true,
          "field": "coreTitleID",
          "headerName": "Title ID",
          "hide": true,
          "javaVariableName": "coreTitleID",
          "searchDataType": "string",
          "width": 100,
        },
        Object {
          "colId": "title",
          "enableSearch": true,
          "field": "title",
          "headerName": "Title",
          "hide": true,
          "javaVariableName": "title",
          "searchDataType": "string",
          "width": 100,
        },
        Object {
          "colId": "series",
          "enableSearch": true,
          "field": "series",
          "headerName": "Series Title",
          "hide": true,
          "javaVariableName": "series",
          "searchDataType": "string",
          "width": 120,
        },
        Object {
          "colId": "seasonNumber",
          "enableSearch": true,
          "field": "seasonNumber",
          "headerName": "Season Number",
          "hide": true,
          "javaVariableName": "seasonNumber",
          "searchDataType": "string",
          "width": 100,
        },
        Object {
          "colId": "episodeNumber",
          "enableSearch": true,
          "field": "episodeNumber",
          "headerName": "Episode Number",
          "hide": true,
          "javaVariableName": "episodeNumber",
          "searchDataType": "string",
          "width": 100,
        },
      ]
    }
    notFilterableColumns={
      Array [
        "action",
      ]
    }
    onGridEvent={[Function]}
    onSortChanged={[Function]}
    rowSelection="multiple"
    setDisplayedRows={[Function]}
    setTotalCount={[Function]}
    suppressRowClickSelection={true}
  />
  <DopTasksTableStatusBar
    paginationData={
      Object {
        "pageSize": 0,
        "totalCount": 0,
      }
    }
  />
</div>
`;
